### DISCUSSION TITLE
Handling of \frametitle and \framesubtitle

### DISCUSSION THREAD

**Discussion by josephwright** [josephwright/ltx-talk#12] [General]
The optional-group syntax in `beamer` is not one we want to carry forward. That leaves us with the question of how to handle frame (sub)titles
```latex
\begin{frame}{mandatory-frame-title}
```
or
```latex
\begin{frame}
  \frametitle{Optional frame title}
```
or something else. Subtitles make this more tricky: it's hard to see they could be mandatory, but they are likely wanted.

**Comment by samcarter** [comment]
Personally, I've always been a fan of the 

```
\begin{frame}
  \frametitle{Optional frame title}
```

syntax. In beamer it made the frametitle a lot more flexible to use, e.g. for overlays. 

-----------------

Just to throw in a different approach for brain-storming: how about key-values in frame options, e.g. `\begin{frame}[fragile,title={Quack},subtitle={Breakfast}]`

**Comment by Skillmon** [comment]
Wanted to throw in the key-values as well.

I don't think that changing the `\frametitle` depending on an overlay is a good idea, that should be an entirely new frame (also in view of tagging in the future this will make things easier).

**Reply by FrankMittelbach** [reply]
I have quite a number of decks where the frame title varies depending on the overlay inside and I don't think that makes it several frames.

But I never used `\frametitle` because of that: I always altered the title argument using `\only` or `\visible` inside, ie the flexibility is there without the need for a separate command

**Reply by u-fischer** [reply]
From the view of accessibility and reading order you have in such a case several frames, simply because you can't in sensible way collapse that into one "handout" frame. E.g. the Question/Answer titles @samcarter's example obviously can only be read if you view them as two frames, regardless about the technical implementation.

**Reply by josephwright** [reply]
Indeed - whilst I'm not keen, I suspect we may end up having to allow more than one `handout` page per `frame` for this reason.  /That said, for e.g. questions and answers, I'd expect them to simply be in one page intermixed, as I'd show the questions 'fill in the blanks' style and simply have the answers in the right places in the handout.

**Reply by FrankMittelbach** [reply]
> From the view of accessibility and reading order you have in such a case several frames, simply because you can't in sensible way collapse that into one "handout" frame.

In my terminology that still remains a single frame (but of course with several slides to put on top of each other).

Overlays come in two forms a) uncovering and b) rewriting+covering. In case of rewrites that certainly poses a question how to  represent that in a different medium and yes in that case you may have to offer slide by slide unless you try to explain  in words what gets covered, what replaced and what added. And I guess we agree that this would be very difficult to do and even if possible probably not "accessible" aka understandable in that case.

So in my opinion, tagging a deck could be done as simple handouts frames when all the overlays do is uncovering and the moment you do more "interactive" stuff the tagging needs changing and perhaps initially simply by representing each overlay slide on its own (which means a lot of  repretition for the consumer, but ...)

**Comment by samcarter** [comment]
> I don't think that changing the `\frametitle` depending on an overlay is a good idea, that should be an entirely new frame (also in view of tagging in the future this will make things easier).

I find things like 

```
\frametitle<1>{Question:}
\frametitle<2>{Answer:}
```

very useful. This avoids a lot of code duplication compared to using two separate frames.

**Comment by josephwright** [comment]
> Just to throw in a different approach for brain-storming: how about key-values in frame options, e.g. `\begin{frame}[fragile,title={Quack},subtitle={Breakfast}]`

I'd thought about that (should have listed it earlier), but I worry

- That for many people this makes an 'optional' argument mandatory
- That it heavily intermixes two very different types of data - typesetting material (titles) and control info (_e.g._ `fragile`)

**Reply by FrankMittelbach** [reply]
My personal approach (with templates) is that the document data that is nearly always necessary should be in positional arguments rather than in key/vals, e.g, it makes totally sense for me to say that a title to a `\section` is a mandatory argument even if the odd heading comes without a title. Same for frames: I would require it as an argument. The less common subframetitle I would put in a key

**Reply by josephwright** [reply]
So you'd go with something that's more like the `beamer` syntax, but with the arg. mandatory - and accept that a reasonable number of frames will be `\begin{frame}{}`? Does that help us with the semantics of the title - is it a sectionning concept?

**Reply by FrankMittelbach** [reply]
My expectation (could be wrong) is that you are the outlyer there --- basically all decks I have seen in the past use frame titles, that is the argument version `\begin{frame{foo}`, because basically nearly all beamer styles reserve space for it so people use it.

Concerning  "what is this title" I guess I confused myself in thinking that `\frametitle` is something different from the arg after `{frame}` --- comes from being half asleep, or old, or ...

New try:  from a tagging perspective I guess it is something like a Caption (like the title of theorems and such) but you could also, I guess, think of it of a heading that is  nested inside any outer `\section` (or `\subsection`) if you have those outside of your frames

**Reply by josephwright** [reply]
> My expectation (could be wrong) is that you are the outlyer there --- basically all decks I have seen in the past use frame titles, that is the argument version `\begin{frame{foo}`, because basically nearly all beamer styles reserve space for it so people use it.

You've seen my slides, right :)

I was actually meaning slides in general, not `beamer` ones - the meeting I was at today/yesterday had I think no LaTeX-based slides, they were all GUI ones - but most of the slides didn't have a title.

> Concerning "what is this title" I guess I confused myself in thinking that `\frametitle` is something different from the arg after `{frame}` --- comes from being half asleep, or old, or ...

Ah, right :)

> New try: from a tagging perspective I guess it is something like a Caption (like the title of theorems and such) but you could also, I guess, think of it of a heading that is nested inside any outer `\section` (or `\subsection`) if you have those outside of your frames

Back with what I suggested to start with :)

**Reply by josephwright** [reply]
If we do decide the frame title is like a theorem name, then does that mean that subtitles don't generally make sense? I'm not seeing much push for subtitles (and we could have say that a split at some pre-defined token in the title denotes a subtitle - if needed).

**Reply by cfr42** [reply]
I do use subtitles, but I wouldn't put much energy into defending them.

**Reply by FrankMittelbach** [reply]
> If we do decide the frame title is like a theorem name, then does that mean that subtitles don't generally make sense? I'm not seeing much push for subtitles (and we could have say that a split at some pre-defined token in the title denotes a subtitle - if needed).

I wouldn't say that they do not make sense; subtitles are just structuring a title. We don't have it but you could imagine that for theorems or section headings etc too. They certainly make sense in the "visual" world and as far as tagging is concerned you could encode that too in some way (or initially just turn the concatenation of title + subtitle into a single data object (loosing some information there  but ...). 

As to the interface on doc level, it could be key/val or it could be some token as you suggest. Not sure which would be a nice one though. Maybe even `\\` but that would then prohibit using this for denoting a line break in a large title (and you would have to resort to `\newline` for that to avoid the subtitle

**Reply by josephwright** [reply]
I was thinking that _if_ we decide that subtitles should be handled within a frame title, _then_ we'd have a dedicated marker (a bit like `\and` in the author list). But at the moment this is all up in the air - we need real presentations with structure to explore.

**Reply by Skillmon** [reply]
I used subtitles extensively in the past, grouping a slew of frames under the same frame title (well, in my case `\section` really), and giving the individual frames different sub titles. Something like

- The Algorithm
    - The idea
    - Basic equations
    - Optimisation criteria and cost functions
    - Extending the algorithm: Kernel functions

So I'm definitely on the pro-subtitles team.

**Comment by Skillmon** [comment]
> > I don't think that changing the `\frametitle` depending on an overlay is a good idea, that should be an entirely new frame (also in view of tagging in the future this will make things easier).
> 
> I find things like
> 
> ```
> \frametitle<1>{Question:}
> \frametitle<2>{Answer:}
> ```
> 
> very useful. This avoids a lot of code duplication compared to using two separate frames.

Don't get me wrong, I didn't mean it couldn't be useful (and I see the appeal), I'm just wary whether the advantages outweigh the implications. That's something you two are certainly better suited to assert.

**Comment by josephwright** [comment]
Looking back, I should have added I do wonder about `fragile` - we could reasonably simply handle all frames that way all of the time (yes, it's slower but possibly the balance here is more toward utility than speed).

**Comment by FrankMittelbach** [comment]
you could have both, i.e., all frames doing fragile handling by default but offer robust-frames=true/false on package load (with default=true) and those concerned about speed (and I have decks with 500+ pages) can set it to false and then need to add "fragile" whenever necessary.

**Comment by cfr42** [comment]
Compilation speed is a huge deal for me. I very rarely need `fragile`.

**Comment by samcarter** [comment]
One point which might be against `\frametitle`: the other approaches might make it easier for xbeamer to know already at the start of the frame if there is going to be a frametitle.

- the frametitle could be set in the header, which would make it easier to create layouts similar to the Berkeley theme
- a more meaningful textheight could be calculated

**Reply by josephwright** [reply]
As we collect the body of the frame, we can always simply do a search up-front for `\frametitle` and the following  _balanced text_, and extract before we typeset the content. So in that sense it's not so important. On the other hand, that's obviously a bit of work we can avoid if we only use an argument to `\begin{frame}`.

**Comment by wspr** [comment]
Since this discussion has become somewhat broader, I thought it would be good to give a couple quick examples from my own course notes.

See first example image below -- the top four lines of the slide are section, subsection, frametitle, framesubtitle. Sections are used to delineate "topics" in the course, and subsections used to essentially mark where I end one video recording and start the next ("subtopics" or "key concepts"). I use the "lecture" subdivision to mark "weeks" (each week contains a small number of topics) but it would make more sense to me to just use \part for the higher level of heading than \section.

Each topic starts with a table of contents -- see second image below. The fact that almost all of this functionality comes "for free" in beamer is essential to preserve in my opinion. I think it would be unfortunate to lose the ability to create collated course notes from individual source files and structure the PDF in an appropriately hierarchical way.

I thought I used subtitles more than I actually do, based on a quick scan over this slide deck. I like their existence -- I don't mind Frank's suggestion to concatenate it with the frame title for tagging. 

<img width="1966" alt="Screenshot 2025-03-20 at 21 09 31" src="https://github.com/user-attachments/assets/077908ed-b155-4178-b89b-0392a81d70bd" />

<img width="1966" alt="Screenshot 2025-03-20 at 21 13 53" src="https://github.com/user-attachments/assets/2049b198-d073-4962-a174-2f9abf62fb2b" />

**Reply by wspr** [reply]
(please excuse the typo in the description of the maths, this is an old version of the slide deck I had on hand)

**Reply by josephwright** [reply]
> Each topic starts with a table of contents -- see second image below. The fact that almost all of this functionality comes "for free" in beamer is essential to preserve in my opinion.

Yes, definitely we need that -  I use a TOC per section myself in longer series of lectures, so it's on the 'draft, that's broken' list at the moment :)

**Reply by josephwright** [reply]
@wspr Very useful to see someone else's teaching content - thanks - I'm guessing the two columns are blocks?

**Reply by josephwright** [reply]
Or is this just two columns with free text?

**Reply by wspr** [reply]
@josephwright  — just `\begin{columns} \column{0.5\linewidth}...` or whatever the syntax is :) I actually haven't used blocks much... something for me to keep in mind!

**Comment by josephwright** [comment]
I've been thinking a bit about frame titles, specifically as they look like we should tag them as Captions. That means that frames look quite like floats (an env with a caption), and that suggests to me that the `\frametitle` approach is likely the best for consistency with the kernel. [Yes, in many ways one with hindsight would do `\begin{figure}{caption}` or similar (cf. ConTeXt), but that ship sailed in the mid 1980s.]

Going with `\frametitle` allows for adding an overlay spec, so there are advantages. We can happily have `\framesubtitle` too here.

So unless there are strong arguments, I think we'll likely stick with the current model.

**Comment by cfr42** [comment]
I use subtitles quite a lot - probably similar to the way other people have described using them. Like most people's, my slides are now on a VLE behind a wall, but I have some older ones which aren't e.g. https://cfrees.wordpress.com/wp-content/uploads/2012/01/2-persons-h.pdf. Note these are all created with `handout` mode. In class, I'd use a version with overlays, but the use of subtitles is the same.
